plugins/experimental/stale_while_revalidate/stale_while_revalidate.c

```diff
@@ -143,47 +116,46 @@ free_request_info(RequestInfo *req_info)
   TSfree(req_info->effective_url);
   TSHandleMLocRelease(req_info->buf, TS_NULL_MLOC, req_info->http_hdr_loc);
   TSMBufferDestroy(req_info->buf);
-  TSfree(req_info->client_addr);
-  TSfree(req_info);
 }
 
 static void
 free_state_info(StateInfo *state)
 {
-  if (state->resp_info) 
-     free_response_info(state->resp_info);
-  if (state->req_info) 
-     free_request_info(state->req_info);
+  if (state->req_info)
+    free_request_info(&state->req_info_obj);
 
   state->req_info  = NULL;
-  state->resp_info = NULL;
 
   TSfree(state);
 }
 
 static CachedHeaderInfo *
-get_cached_header_info(TSHttpTxn txn)
+get_cached_header_info(CachedHeaderInfo *chi, TSHttpTxn txn)
 {
-  CachedHeaderInfo *chi;
   TSMBuffer cr_buf;
   TSMLoc cr_hdr_loc, cr_date_loc, cr_cache_control_loc, cr_cache_control_dup_loc;
   int cr_cache_control_count, val_len, i;
   char *value, *ptr;
 
-  chi                         = (CachedHeaderInfo *)TSmalloc(sizeof(CachedHeaderInfo));
-  chi->date                   = 0;
+  chi->date                   = 0; // internal
+  chi->rmt_date               = 0; // cache header
   chi->max_age                = 0;
   chi->stale_while_revalidate = 0;
   chi->stale_on_error         = 0;
 
+  TSHttpTxnCachedRespTimeGet(txn, &chi->date); // show local-clock age
+
   if (TSHttpTxnCachedRespGet(txn, &cr_buf, &cr_hdr_loc) == TS_SUCCESS) {
     cr_date_loc = TSMimeHdrFieldFind(cr_buf, cr_hdr_loc, TS_MIME_FIELD_DATE, TS_MIME_LEN_DATE);
     if (cr_date_loc != TS_NULL_MLOC) {
       TSDebug(PLUGIN_NAME, "Found a date");
-      chi->date = TSMimeHdrFieldValueDateGet(cr_buf, cr_hdr_loc, cr_date_loc);
+      chi->rmt_date = TSMimeHdrFieldValueDateGet(cr_buf, cr_hdr_loc, cr_date_loc);
```

Comment:
remote clocks can be off versus the (obviously) superior and foolproof clock of the local system...